cmd/roles/roles.go

```diff
@@ -415,6 +415,35 @@ func (mr *MilvusRoles) Run() {
 		componentMap[typeutil.QueryCoordRole] = queryCoord
 	}
 
+	waitCoordBecomeHealthy := func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Info("wait all coord become healthy loop quit")
+				return
+			default:
+				rcState := rootCoord.Health(ctx)
+				dcState := dataCoord.Health(ctx)
+				icState := indexCoord.Health(ctx)
+				qcState := queryCoord.Health(ctx)
+
+				if rcState == commonpb.StateCode_Healthy && dcState == commonpb.StateCode_Healthy && icState == commonpb.StateCode_Healthy && qcState == commonpb.StateCode_Healthy {
+					log.Info("all coord become healthy")
+					return
+				}
+				log.Info("wait all coord become healthy", zap.String("rootCoord", rcState.String()), zap.String("dataCoord", dcState.String()), zap.String("indexCoord", icState.String()), zap.String("queryCoord", qcState.String()))
+				time.Sleep(time.Second)
+			}
+		}
+	}
+
+	// In standalone mode, block the start process until the new coordinator is active to avoid the coexistence of the old coordinator and the new node/proxy
+	// 1. In the start/restart process, the new coordinator will become active immediately and will not be blocked
+	// 2. In the rolling upgrade process, the new coordinator will not be active until the old coordinator is down, and it will be blocked
+	if mr.Local {
+		waitCoordBecomeHealthy()
```

Comment:
I don't think this is gonna to enough.

For all the standalone node, even it's get healthy it doesn't mean they become leader.

I think we should add check here


	// stop coordinators first
	coordinators := []component{rootCoord, dataCoord, indexCoord, queryCoord}
	for idx, coord := range coordinators {
		log.Warn("stop processing cooordinator xxx")
		if coord != nil {
			log.Info("stop coord", zap.Int("idx", idx), zap.Any("coord", coord))
			coord.Stop()
		}
                 log.Info("start to wait cooordinator xxx becomes leader")
                 // ADD SOME BARRIER HERE
                waitNewCoordinatorBecomesActive()
                log.Info("succesfully wait cooordinator xxx becomes leader")
	}
	log.Info("All coordinators have stopped")

	// stop nodes
	nodes := []component{queryNode, indexNode, dataNode}
	for idx, node := range nodes {
                log.Warn("start to stop querynode xxx")
		if node != nil {
			log.Info("stop node", zap.Int("idx", idx), zap.Any("node", node))
			node.Stop()
		}
                log.Info("start to drain all data from Querynode/Datanode")
                 // ADD SOME BARRIER HERE
                waitToDrainData()
                log.Info("succesfully drain datanode/querynode")
	}
	log.Info("All nodes have stopped")

Reply:
when coord is standby, it's in `standby` state, and after it become active coord, it state will change to `healthy`.

so if we block the new node/proxy's online, we can avoid the coexistence.

but if we block the stop progress, which means the `restart/stop` progress will be stuck too